### PR TITLE
Merge extension with extended type

### PR DESCRIPTION
**Pull request by pcantrell**
Continuing #289 based off of master instead of swift-2.0.

### CODE REVIEW

**Comment by pcantrell** [comment]
I went ahead and got this one ready, now that @jpsim big Obj-C push is done. It’s now based off of master instead of swift-2.0, and the specs are up to date.

This is ready to merge if you’re happy with the output. As a reminder of issues from the older discussion:

- @jpsim was concerned about clobbering `// MARK:` text with the “Default Implementation” heading. That’s now fixed.
- There was concern about #333, which turned out not to be related to this PR.
- #334 limits the usefulness of this feature somewhat, but again, that problem existed before this PR.

File: lib/jazzy/source_declaration/type.rb [hunk 1]
```diff
@@ -61,6 +61,14 @@ def extension?
         kind =~ /^source\.lang\.swift\.decl\.extension.*/
       end
 
+      def extensible?
+        kind =~ /^source\.lang\.swift\.decl\.(class|struct|protocol|enum)$/
+      end
+
+      def protocol?
+        kind =~ /^source\.lang\.swift\.decl\.protocol$/
```
Comment (jpsim) [inline lib/jazzy/source_declaration/type.rb]: it seems nicer/cleaner to just do a string comparison if we're matching for an exact string anyway?

File: lib/jazzy/source_declaration/type.rb [hunk 2]
```diff
@@ -61,6 +61,14 @@ def extension?
         kind =~ /^source\.lang\.swift\.decl\.extension.*/
       end
 
+      def extensible?
```
Comment (jpsim) [inline lib/jazzy/source_declaration/type.rb]: Now that we have both ObjC & Swift types in there, I've prefixed the other language-specific methods with `objc_` and `swift_`. Could you please do the same for `extensible?` and `protocol?`.

**Comment by jpsim** [comment]
This is so so great. I'm still impressed at how concise you managed to keep this code.

:+1: other than my very minor comments.

Also, although this is something I haven't done yet, there are lots of areas now where Swift processing differs from Objective-C processing, and we should at some point go through the codepaths to branch off where needed to keep things efficient.

File: lib/jazzy/source_declaration/type.rb (hunk 2 above)
Comment (segiddins) [inline lib/jazzy/source_declaration/type.rb]: I don't see why we wouldn't want this new behavior to also apply in ObjC?

Comment (jpsim) [inline lib/jazzy/source_declaration/type.rb]: I feel like Objective-C categories aren't as *lightweight* as Swift extensions. You name your Objective-C categories. Appledoc also treats them as different than their extending type.

Comment (pcantrell) [inline lib/jazzy/source_declaration/type.rb]: Updating with `swift_` and `objc_` prefixes as you describe for this PR.

What do you think about generic tests like `protocol?` and `extensible?` plus orthogonal language tests like `objc?` and `swift?`?

**Comment by pcantrell** [comment]
>This is so so great. I'm still impressed at how concise you managed to keep this code.

That’s kind of you. I actually found it a little hard to read on re-review, so I did a very light refactoring that I think helps.

@jpsim, I’ve made your very sensible fixes, so if CI likes it, we’re good to go.

File: lib/jazzy/source_declaration/type.rb (hunk 2 above)
Comment (jpsim) [inline lib/jazzy/source_declaration/type.rb]: > What do you think about refactoring to generic tests like protocol? and extensible? plus orthogonal language tests like objc? and swift? (after this snowballing PR is merged)?

I think that's sensible as long as they're used by both languages.

**Comment by jpsim** [comment]
Looks like there are rubocop violations to address:

```
Offenses:
lib/jazzy/sourcekitten.rb:313:9: C: Align the operands of an expression in an assignment spanning multiple lines.
        .group_by { |d| deduplication_key(d) }
        ^^^^^^^^^
lib/jazzy/sourcekitten.rb:314:9: C: Align the operands of an expression in an assignment spanning multiple lines.
        .values
        ^^^^^^^
22 files inspected, 2 offenses detected
RuboCop failed!
The command "bundle exec rake spec" exited with 1.
```

**Comment by pcantrell** [comment]
Yeah, weird — why didn’t those show up locally? Fixing…

**Comment by pcantrell** [comment]
Not getting that rubocop failure locally. I notice that I have rubocop 0.26.1 locally, which is what’s in the Gemfile.lock, but Travis is downloading 0.34.2. ?!?

Just pushed a commit that’s a shot in the dark at what I _think_ rubocop is asking for. We’ll see if it passes.

**Comment by jpsim** [comment]
Rubocop is pinned at `0.34.2` in the Gemfile.lock on master: https://github.com/realm/jazzy/blob/master/Gemfile.lock#L88

In any case, your shot in the dark passed! :tada: 

I'll add a changelog entry for you in the spirit of getting this in.

**Comment by pcantrell** [comment]
GTG

**Comment by pcantrell** [comment]
Oh, right , changelog. Thanks.

**Pull request merged**